src/pow.rs

```diff
@@ -0,0 +1,1314 @@
+            ///
+            /// ```
+            #[doc = concat!("use bitcoin::", stringify!($ty), ";")]
+            #[doc = concat!("let a = ", stringify!($ty), "::from(0xdeadbeef);")]
+            /// assert_eq!(a.to_le(), [0xdeadbeef, 0, 0, 0]);
+            /// ```
+            #[inline]
+            pub fn to_le(self) -> [u64; 4] { self.0.to_le() }
+
+            /// Converts `self` to a big-endian byte array.
+            #[inline]
+            pub fn to_be_bytes(self) -> [u8; 32] { self.0.to_be_bytes() }
+
+            /// Converts `self` to a little-endian byte array.
+            #[inline]
+            pub fn to_le_bytes(self) -> [u8; 32] { self.0.to_le_bytes() }
+        }
+
+        impl From<u64> for $ty {
+            #[inline]
+            fn from(x: u64) -> Self { $ty(Uint256::from(x)) }
+        }
+
+        impl fmt::Display for $ty {
+            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result { fmt::Display::fmt(&self.0, f) }
+        }
+
+        impl Encodable for $ty {
+            #[inline]
+            fn consensus_encode<W: crate::io::Write + ?Sized>(
+                &self,
+                w: &mut W,
+            ) -> Result<usize, crate::io::Error> {
+                self.0.consensus_encode(w)
+            }
+        }
+
+        impl Decodable for $ty {
+            fn consensus_decode<R: crate::io::Read + ?Sized>(
+                r: &mut R,
+            ) -> Result<$ty, crate::consensus::encode::Error> {
+                let u: Uint256 = Decodable::consensus_decode(r)?;
+                Ok($ty(u))
+            }
+        }
+    };
+}
+
+/// Implements `ops::Mul` by integer type in both directions. Requires `u32::from($int)`.
+macro_rules! impl_mul {
+    ($ty:ident, $int:ident) => {
+        impl Mul<$int> for $ty {
+            type Output = $ty;
+
+            #[inline]
+            fn mul(self, rhs: $int) -> Self { $ty(self.0.mul_u32(rhs.into())) }
+        }
+
+        impl Mul<$ty> for $int {
+            type Output = $ty;
+
+            #[inline]
+            fn mul(self, rhs: $ty) -> $ty { $ty(rhs.0.mul_u32(self.into())) }
+        }
+    };
+}
+
+/// Implements `ops::Div` by integer type `$ty`. Requires `u64::from($ty)`.
+macro_rules! impl_div {
+    ($ty:ident, $int:ident) => {
+        impl core::ops::Div<$int> for $ty {
+            type Output = $ty;
+
+            #[inline]
+            fn div(self, rhs: $int) -> Self {
+                let rhs = Uint256::from(u64::from(rhs));
+                $ty(self.0 / rhs)
+            }
+        }
+    };
+}
+
+/// Implements `ops::Shr` and `ops::Shl` by integer type `$ty`. Silently casts shift argument to
+/// `usize`, this is safe as long as `$ty` never has more bits than the underlying architecture.
+macro_rules! impl_shift {
+    ($ty:ident, $int:ident) => {
+        impl Shl<$int> for $ty {
+            type Output = Self;
+
+            #[inline]
+            fn shl(self, shift: $int) -> Self { Self(self.0 << shift as usize) }
```

Comment:
This is weird, shifts in `std` accept `u32` not `usize`

Reply:
Yeah I thought as much but everytime I try to use `u32` I get in a knot of casting types, and removing casts, all over the place. I'll have another go now you've prompted me.

Reply:
Fixed, all use `u32` now.